Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,835 @@
+            // further frames as more of the payload arrives
+            let streamID = HTTP2StreamID(knownID: Int32(header.rawStreamID & ~0x8000_0000))
+            guard streamID != .rootStream else {
+                // DATA frames MUST be associated with a stream. If a DATA frame is received whose
+                // stream identifier field is 0x0, the recipient MUST respond with a connection error
+                // (Section 5.4.1) of type PROTOCOL_ERROR.
+                throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+            }
+            
+            var slice = bytes
+            let consumed = slice.readableBytes
+            bytes.clear()       // consumed all the bytes
+            
+            let flags = HTTP2Frame.FrameFlags(rawValue: header.flags)
+            let padding: UInt8
+            if flags.contains(.padded) {
+                padding = slice.readInteger()!
+            } else {
+                padding = 0
+            }
+            self.state = .simulatingDataFrames(originalHeader: header, padding: padding, remainingBytes: header.length - consumed)
+            
+            // NB: our synthetic frames never claim padding, nor are they (by definition) a stream end
+            return HTTP2Frame(streamID: streamID, flags: [], payload: .data(.byteBuffer(slice)))
+        }
+    }
+    
+    private mutating func readFrame(withHeader header: FrameHeader, from bytes: inout ByteBuffer) throws -> HTTP2Frame? {
+        assert(bytes.readableBytes >= header.length, "Buffer should contain at least \(header.length) bytes.")
+        
+        let flags = HTTP2Frame.FrameFlags(rawValue: header.flags)
+        let streamID = HTTP2StreamID(knownID: Int32(header.rawStreamID & ~0x8000_0000))
+        
+        // Whatever else happens (e.g. errors, padding), we want to ensure that the input buffer
+        // has had all this frame's bytes consumed when we exit. This also allows us to ignore
+        // trailing padding bytes in the per-frame decoders below, and to ignore the status of
+        // the input buffer if we pull out a slice on which to operate: this will ensure the right
+        // output state of the input buffer.
+        let frameEndIndex = bytes.readerIndex + header.length
+        defer {
+            bytes.moveReaderIndex(to: frameEndIndex)
+        }
+        
+        let payload: HTTP2Frame.FramePayload
+        do {
+            switch header.type {
+            case 0:
+                payload = try self.parseDataFramePayload(length: header.length, streamID: streamID, flags: flags, bytes: &bytes)
+            case 1:
+                if flags.contains(.endHeaders) {
+                    payload = try self.parseHeadersFramePayload(length: header.length, streamID: streamID, flags: flags, bytes: &bytes)
+                } else {
+                    let slice = bytes.readSlice(length: header.length)!
```

Comment:
This mishandles the possibility that this frame contains PADDING: we need to strip that off.

Reply:
This has all been rewritten with a new state machine.